Equipe Carlos Filipe, Celso e Charlene

Importações
<br>Origem: Kaggle
<br>Base: Campeonato Brasileiro de FutebolImportações:

In [2]:
# Instalar para permitir o carregamendo dos dados:
# pip install kagglehub[pandas-datasets]
import kagglehub
from kagglehub import KaggleDatasetAdapter

# Definindo nomes dos datasets que serão carregados
estatisticas = "campeonato-brasileiro-estatisticas-full.csv"
full = "campeonato-brasileiro-full.csv"

# Carregando dataset campeonato-brasileiro-de-futebol Estatísticas
df_estats = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "adaoduque/campeonato-brasileiro-de-futebol",
  estatisticas,

)

# Carregando dataset campeonato-brasileiro-de-futebol Full
df_full = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "adaoduque/campeonato-brasileiro-de-futebol",
  full,
)

# Exibindo Head, Shape e Info dos datasets carregados
display(df_estats.head())
display(df_full.head())

print(f"Shape do df_estats: {df_estats.shape}")
print(f"Shape do df_full: {df_full.shape}")

df_estats.info()
df_full.info()

c:\Users\Praia\AppData\Local\Programs\Python\Python314\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
C:\Users\Praia\AppData\Local\Temp\ipykernel_558988\2682929807.py:11: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  df_estats = kagglehub.load_dataset(
C:\Users\Praia\AppData\Local\Temp\ipykernel_558988\2682929807.py:19: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  df_full = kagglehub.load_dataset(


,partida_id,rodata,clube,chutes,chutes_no_alvo,posse_de_bola,passes,precisao_passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios
0,1,1,Vasco,0,0,NaN,0,NaN,0,0,0,0,0
1,1,1,Guarani,0,0,NaN,0,NaN,0,0,0,0,0
2,2,1,Athletico-PR,0,0,NaN,0,NaN,0,0,0,0,0
3,2,1,Gremio,0,0,NaN,0,NaN,0,0,0,0,0
4,3,1,Flamengo,0,0,NaN,0,NaN,0,0,0,0,0


,ID,rodata,data,hora,mandante,visitante,formacao_mandante,formacao_visitante,tecnico_mandante,tecnico_visitante,vencedor,arena,mandante_Placar,visitante_Placar,mandante_Estado,visitante_Estado,arrecadacao
0,1,1,29/03/2003,16:00,Guarani,Vasco,NaN,NaN,NaN,NaN,Guarani,Brinco de Ouro,4,2,SP,RJ,NaN
1,2,1,29/03/2003,16:00,Athletico-PR,Gremio,NaN,NaN,NaN,NaN,Athletico-PR,Arena da Baixada,2,0,PR,RS,NaN
2,3,1,30/03/2003,16:00,Flamengo,Coritiba,NaN,NaN,NaN,NaN,-,Maracanã,1,1,RJ,PR,NaN
3,4,1,30/03/2003,16:00,Goias,Paysandu,NaN,NaN,NaN,NaN,-,Serra Dourada,2,2,GO,PA,NaN
4,5,1,30/03/2003,16:00,Internacional,Ponte Preta,NaN,NaN,NaN,NaN,-,Beira Rio,1,1,RS,SP,NaN


Shape do df_estats: (18330, 13)
Shape do df_full: (9165, 17)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18330 entries, 0 to 18329
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   partida_id       18330 non-null  int64 
 1   rodata           18330 non-null  int64 
 2   clube            18330 non-null  object
 3   chutes           18330 non-null  int64 
 4   chutes_no_alvo   18330 non-null  int64 
 5   posse_de_bola    7568 non-null   object
 6   passes           18330 non-null  int64 
 7   precisao_passes  5264 non-null   object
 8   faltas           18330 non-null  int64 
 9   cartao_amarelo   18330 non-null  int64 
 10  cartao_vermelho  18330 non-null  int64 
 11  impedimentos     18330 non-null  int64 
 12  escanteios       18330 non-null  int64 
dtypes: int64(10), object(3)
memory usage: 1.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9165 entries, 0 to 9164
Data columns (total 17 columns

<font size='5'>1 Descrição inicial</font>

<br>O que cada coluna significa?
<br>Qual o recorte temporal?
<br>3 a 5 perguntas** que vocês querem responder, definidas **antes** da análise

Abaixo realizando o merge de dados da tabela Full na tabela Estatísticas relacionando a coluna ID da Partida.

In [3]:
df_merge = df_estats.merge(
    df_full, 
    left_on='partida_id',   # Chave no primeiro dataset
    right_on='ID',          # Chave no segundo dataset
    how='inner'              
)

print(df_merge)

       partida_id  rodata_x          clube  chutes  chutes_no_alvo  \
0               1         1          Vasco       0               0   
1               1         1        Guarani       0               0   
2               2         1   Athletico-PR       0               0   
3               2         1         Gremio       0               0   
4               3         1       Flamengo       0               0   
...           ...       ...            ...     ...             ...   
18325        9163        38      Fortaleza      10               5   
18326        9164        38    Atletico-MG      19               5   
18327        9164        38          Vasco       1               0   
18328        9165        38     Bragantino       5               3   
18329        9165        38  Internacional      14               4   

      posse_de_bola  passes precisao_passes  faltas  cartao_amarelo  ...  \
0               NaN       0             NaN       0               0  ...   
1      

Recorte temporal
<br>
<br> Selecionados dados de 2018 a 2023.
<br> Até 2017 e a partir de 2024 muitos dados estão ausentes e prejudicam a análise temporal.

In [4]:
import pandas as pd

# Converte a coluna 'data' para datetime
df_merge['data'] = pd.to_datetime(df_merge['data'], format='%d/%m/%Y')

# Filtra os anos de 2018 e 2023 - até 2017 e a partir de 2024 muitos dados estão ausentes e prejudicam a análise temporal.
df_filtrado = df_merge[(df_merge['data'].dt.year >= 2018) & (df_merge['data'].dt.year <= 2023)]

# Renomeando coluna com nome autonomeado no merge
df_filtrado = df_filtrado.rename(columns={'rodata_x': 'rodada'})
df_filtrado.describe()

,partida_id,rodada,chutes,chutes_no_alvo,passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,ID,rodata_y,data,mandante_Placar,visitante_Placar,arrecadacao
count,4560.000000,4560.000000,4560.000000,4560.000000,4560.000000,4560.000000,4560.000000,4560.000000,4560.00000,4560.000000,4560.000000,4560.000000,4560,4560.000000,4560.000000,0.0
mean,7265.500000,19.500000,12.965132,4.236842,420.145614,14.615351,2.291667,0.131798,1.47500,5.203289,7265.500000,19.500000,2021-03-04 19:51:09.473684224,1.378070,0.964035,NaN
min,6126.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,6126.000000,1.000000,2018-04-14 00:00:00,0.000000,0.000000,NaN
25%,6695.750000,10.000000,9.000000,3.000000,347.000000,12.000000,1.000000,0.000000,0.00000,3.000000,6695.750000,10.000000,2019-09-19 12:00:00,1.000000,0.000000,NaN
50%,7265.500000,19.500000,12.500000,4.000000,418.000000,14.000000,2.000000,0.000000,1.00000,5.000000,7265.500000,19.500000,2021-04-12 12:00:00,1.000000,1.000000,NaN
75%,7835.250000,29.000000,16.000000,6.000000,492.000000,17.000000,3.000000,0.000000,2.00000,7.000000,7835.250000,29.000000,2022-07-26 06:00:00,2.000000,1.000000,NaN
max,8405.000000,38.000000,37.000000,18.000000,848.000000,31.000000,10.000000,3.000000,9.00000,23.000000,8405.000000,38.000000,2023-12-06 00:00:00,7.000000,6.000000,NaN
std,658.251424,10.967059,5.276672,2.366516,110.956524,4.568393,1.446576,0.373441,1.37817,3.031477,658.251424,10.967059,NaN,1.150279,0.984426,NaN


In [5]:
#Eliminando colunas repetidas ou que não tem relevância na análise e criando o dataframe que será usado nos tratamentos de dados daqui pra frente
df_diagnostico = df_filtrado.drop(columns=['ID','rodata_y','tecnico_mandante','tecnico_visitante','arena','arrecadacao'])

<font size='5'>2 Diagnóstico de qualidade</font>

In [6]:
#Sugestão do professor como resumo do diagnóstico
def diagnostico(df):
    return pd.DataFrame({
        'tipo': df.dtypes.astype(str),
        'faltantes': df.isna().sum(),
        'pct_faltantes': (df.isna().mean() * 100).round(2),
        'distintos': df.nunique(),
        'mais_frequente': [df[c].mode().iloc[0] if df[c].notna().any() else None
                           for c in df.columns],
    })

diagnostico(df_diagnostico)

,tipo,faltantes,pct_faltantes,distintos,mais_frequente
partida_id,int64,0,0.00,2280,6126
rodada,int64,0,0.00,38,1
clube,object,0,0.00,29,Athletico-PR
chutes,int64,0,0.00,37,11
chutes_no_alvo,int64,0,0.00,17,3
posse_de_bola,object,48,1.05,61,47%
passes,int64,0,0.00,508,0
precisao_passes,object,48,1.05,42,82%
faltas,int64,0,0.00,30,14
cartao_amarelo,int64,0,0.00,10,2


In [7]:
# Dimensões, tipos, uso de memória
print("shape:", df_diagnostico.shape)
df_diagnostico.info(memory_usage='deep')

shape: (4560, 24)
<class 'pandas.core.frame.DataFrame'>
Index: 4560 entries, 12250 to 16809
Data columns (total 24 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   partida_id          4560 non-null   int64         
 1   rodada              4560 non-null   int64         
 2   clube               4560 non-null   object        
 3   chutes              4560 non-null   int64         
 4   chutes_no_alvo      4560 non-null   int64         
 5   posse_de_bola       4512 non-null   object        
 6   passes              4560 non-null   int64         
 7   precisao_passes     4512 non-null   object        
 8   faltas              4560 non-null   int64         
 9   cartao_amarelo      4560 non-null   int64         
 10  cartao_vermelho     4560 non-null   int64         
 11  impedimentos        4560 non-null   int64         
 12  escanteios          4560 non-null   int64         
 13  data                4560 non-n

Dimensões e tipos
<br><br>A tabela possui 4560 linhas e 23 colunas sendo
<br>12 Inteiros
<br>1 Datetime
<br>8 Texto
<br>2 Colunas que deveriam ser numéricas para obter estatísticas: posse_de_bola, precisao_passes.
<br>
<br>Uso de memória 3.1 MB

In [8]:
# Estatísticas descritivas das colunas numéricas e texto/objeto

display(df_diagnostico.describe().round(2))
display(df_diagnostico.describe(include='object'))

,partida_id,rodada,chutes,chutes_no_alvo,passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,data,mandante_Placar,visitante_Placar
count,4560.00,4560.00,4560.00,4560.00,4560.00,4560.00,4560.00,4560.00,4560.00,4560.00,4560,4560.00,4560.00
mean,7265.50,19.50,12.97,4.24,420.15,14.62,2.29,0.13,1.48,5.20,2021-03-04 19:51:09.473684224,1.38,0.96
min,6126.00,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,2018-04-14 00:00:00,0.00,0.00
25%,6695.75,10.00,9.00,3.00,347.00,12.00,1.00,0.00,0.00,3.00,2019-09-19 12:00:00,1.00,0.00
50%,7265.50,19.50,12.50,4.00,418.00,14.00,2.00,0.00,1.00,5.00,2021-04-12 12:00:00,1.00,1.00
75%,7835.25,29.00,16.00,6.00,492.00,17.00,3.00,0.00,2.00,7.00,2022-07-26 06:00:00,2.00,1.00
max,8405.00,38.00,37.00,18.00,848.00,31.00,10.00,3.00,9.00,23.00,2023-12-06 00:00:00,7.00,6.00
std,658.25,10.97,5.28,2.37,110.96,4.57,1.45,0.37,1.38,3.03,NaN,1.15,0.98


,clube,posse_de_bola,precisao_passes,hora,mandante,visitante,formacao_mandante,formacao_visitante,vencedor,mandante_Estado,visitante_Estado
count,4560,4512,4512,4560,4560,4560,4560,4560,4560,4560,4560
unique,29,61,42,138,29,29,21,21,30,12,12
top,Santos,53%,82%,19:00,Santos,Flamengo,4-2-3-1,4-2-3-1,-,SP,SP
freq,228,167,357,534,228,228,2356,2208,1270,1064,1064


In [9]:
#jogos com passes zerado
pd.set_option('display.max_columns', None)
display(df_diagnostico[df_diagnostico['passes'] == 0])

,partida_id,rodada,clube,chutes,chutes_no_alvo,posse_de_bola,passes,precisao_passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,data,hora,mandante,visitante,formacao_mandante,formacao_visitante,vencedor,mandante_Placar,visitante_Placar,mandante_Estado,visitante_Estado
13030,6517,2,CSA,0,0,NaN,0,NaN,0,0,0,0,0,2019-05-01,16:00,CSA,Palmeiras,4-2-3-1,4-2-3-1,-,1,1,AL,SP
13031,6517,2,Palmeiras,0,0,NaN,0,NaN,0,0,0,0,0,2019-05-01,16:00,CSA,Palmeiras,4-2-3-1,4-2-3-1,-,1,1,AL,SP
13076,6541,4,Atletico-MG,0,0,NaN,0,NaN,0,0,0,0,0,2019-05-12,10:53,Atletico-MG,Palmeiras,4-3-3,4-2-3-1,Palmeiras,0,2,MG,SP
13077,6541,4,Palmeiras,0,0,NaN,0,NaN,0,0,0,0,0,2019-05-12,10:53,Atletico-MG,Palmeiras,4-3-3,4-2-3-1,Palmeiras,0,2,MG,SP
13178,6595,9,Athletico-PR,0,0,NaN,0,NaN,0,0,0,0,0,2019-06-13,14:08,Goias,Athletico-PR,4-2-3-1,4-1-2-1-2,Goias,2,1,GO,PR
13179,6595,9,Goias,0,0,NaN,0,NaN,0,0,0,0,0,2019-06-13,14:08,Goias,Athletico-PR,4-2-3-1,4-1-2-1-2,Goias,2,1,GO,PR
13210,6609,11,Athletico-PR,0,0,NaN,0,NaN,0,0,0,0,0,2019-07-20,16:10,CSA,Athletico-PR,4-3-1-2,4-2-3-1,Athletico-PR,0,4,AL,PR
13211,6609,11,CSA,0,0,NaN,0,NaN,0,0,0,0,0,2019-07-20,16:10,CSA,Athletico-PR,4-3-1-2,4-2-3-1,Athletico-PR,0,4,AL,PR
13228,6617,12,Cruzeiro,0,0,NaN,0,NaN,0,0,0,0,0,2019-07-27,13:58,Cruzeiro,Athletico-PR,4-2-3-1,4-2-3-1,Athletico-PR,0,2,MG,PR
13229,6617,12,Athletico-PR,0,0,NaN,0,NaN,0,0,0,0,0,2019-07-27,13:58,Cruzeiro,Athletico-PR,4-2-3-1,4-2-3-1,Athletico-PR,0,2,MG,PR


In [10]:
# identificar o que é um NA nessa tabela
# Quantidade e percentual de valores faltantes
faltantes = pd.DataFrame({
    'faltantes': df_diagnostico.isna().sum(),
    'pct': (df_diagnostico.isna().mean() * 100).round(2)
})
faltantes = faltantes[faltantes['faltantes'] > 0].sort_values('faltantes', ascending=False)
faltantes

,faltantes,pct
posse_de_bola,48,1.05
precisao_passes,48,1.05


In [11]:
#Linhas com pelo menos um NA x Linhas completas
print("linhas com algum NaN:", df_diagnostico.isna().any(axis=1).sum())
print("linhas completas    :", df_diagnostico.notna().all(axis=1).sum())

linhas com algum NaN: 48
linhas completas    : 4512


Faltantes por coluna (quantidade e percentual)
<br><br>
Há cerca de 1% de faltantes nas colunas posse_de_bola e precisao_passes.
Linhas para tratamento são as que possuem NaN em posse_de_bola e precisao_passes. Colunas para realizar tratamento: chutes, chutes_no_alvo, posse_de_bola, passes, precisao_passes, faltas, cartao_amarelo, cartao_vermelho, impedimentos e escanteios.

In [12]:
total_duplicadas = df_diagnostico.duplicated().sum()
print(total_duplicadas)

0


Não existem linhas perfeitamente duplicadas.

In [13]:
# Cria a coluna ano para realizar diagnóstico. A coluna deve ser utilizada também na etapa de análise dos dados.
df_diagnostico["ano"] = pd.to_datetime(df_diagnostico["data"]).dt.year
jogos_por_ano = df_diagnostico.groupby("ano").size().reset_index(name="quantidade_de_linhas")

print(jogos_por_ano)

    ano  quantidade_de_linhas
0  2018                   760
1  2019                   760
2  2020                   536
3  2021                   984
4  2022                   760
5  2023                   760


A única coluna numérica que aparenta ter valores inconsistentes é a coluna passes.
O único valor que foge ao padrão é o de 30 clubes vencedores, mas é porque considera o valor "-" que representa o empate.
Não há linhas duplicadas.

Existe uma inconsistência que necessita ser tratada no número de jogos nos anos de 2020 e 2021. Parte dos jogos do campeonato de 2020 foram realizados em 2021 devido à pandemia de covid-19 nos mêses de Janeiro/2021 e Fevereiro/2021.
<br>O campeonato de 2021 começou efetivamente em Maio/2021.

In [14]:
# Para colunas categóricas: Quantos valores distintos e frequência de cada valor

for col in ['clube']:
    print(f"--- {col} | {df_diagnostico[col].nunique()} valores distintos ---")
    print(df_diagnostico[col].value_counts(dropna=False).sort_index(), '\n')

for col in ['visitante']:
    print(f"--- {col} | {df_diagnostico[col].nunique()} valores distintos ---")
    print(df_diagnostico[col].value_counts(dropna=False).sort_index(), '\n')

for col in ['mandante']:
    print(f"--- {col} | {df_diagnostico[col].nunique()} valores distintos ---")
    print(df_diagnostico[col].value_counts(dropna=False).sort_index(), '\n')


--- clube | 29 valores distintos ---
clube
America-MG       152
Athletico-PR     228
Atletico-GO      114
Atletico-MG      228
Avai              76
Bahia            190
Botafogo-RJ      190
Bragantino       152
CSA               38
Ceara            190
Chapecoense      114
Corinthians      228
Coritiba         114
Cruzeiro         114
Cuiaba           114
Flamengo         228
Fluminense       228
Fortaleza        190
Goias            152
Gremio           190
Internacional    228
Juventude         76
Palmeiras        228
Parana            38
Santos           228
Sao Paulo        228
Sport            114
Vasco            152
Vitoria           38
Name: count, dtype: int64 

--- visitante | 29 valores distintos ---
visitante
America-MG       152
Athletico-PR     228
Atletico-GO      114
Atletico-MG      228
Avai              76
Bahia            190
Botafogo-RJ      190
Bragantino       152
CSA               38
Ceara            190
Chapecoense      114
Corinthians      228
Coritiba         

Não foram identificadas inconsistências em colunas categóricas.

In [15]:
#Nas colunas ID identificar se existem linhas duplicadas
print("ids duplicados:", df_diagnostico['partida_id'].duplicated().sum())
print("ids únicos:", df_diagnostico['partida_id'].nunique(), "| linhas:", len(df_diagnostico))

print("Combinações duplicadas de partida_id e clube:", df_diagnostico.duplicated(subset=['partida_id', 'clube']).sum())
print("Combinações únicas:", len(df_diagnostico[['partida_id', 'clube']].drop_duplicates()), "| Total de linhas:", len(df_diagnostico))


ids duplicados: 2280
ids únicos: 2280 | linhas: 4560
Combinações duplicadas de partida_id e clube: 0
Combinações únicas: 4560 | Total de linhas: 4560


Todos os IDs de partida estão duplicados. Cada ID de partida é exibido em duas linhas sendo cada uma com as estatística referentes a cada clube na partida.

In [16]:
#Caso existam colunas total NA, excluir

vazias = df_diagnostico.columns[df_diagnostico.isna().all()]
print("removendo:", list(vazias))
df_diagnostico = df_diagnostico.drop(columns=vazias)
print(df_diagnostico.shape)

removendo: []
(4560, 25)


Não existem colunas total NA.

<font size='5'>3 Limpeza e transformação</font>

In [17]:
df_limpeza = df_diagnostico

#Tranformando posse_de_bola, precisao_passes em colunas numéricas

df_limpeza['posse_de_bola'] = df_limpeza['posse_de_bola'].astype(str).str.replace('%', '').str.strip().astype(float)
df_limpeza['precisao_passes'] = df_limpeza['precisao_passes'].astype(str).str.replace('%', '').str.strip().astype(float)

In [18]:
display(df_limpeza.describe().round(4))

,partida_id,rodada,chutes,chutes_no_alvo,posse_de_bola,passes,precisao_passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,data,mandante_Placar,visitante_Placar,ano
count,4560.0000,4560.0000,4560.0000,4560.0000,4512.0000,4560.0000,4512.0000,4560.0000,4560.0000,4560.0000,4560.0000,4560.0000,4560,4560.0000,4560.0000,4560.0000
mean,7265.5000,19.5000,12.9651,4.2368,50.0000,420.1456,80.4337,14.6154,2.2917,0.1318,1.4750,5.2033,2021-03-04 19:51:09.473684224,1.3781,0.9640,2020.5491
min,6126.0000,1.0000,0.0000,0.0000,20.0000,0.0000,39.0000,0.0000,0.0000,0.0000,0.0000,0.0000,2018-04-14 00:00:00,0.0000,0.0000,2018.0000
25%,6695.7500,10.0000,9.0000,3.0000,42.0000,347.0000,77.0000,12.0000,1.0000,0.0000,0.0000,3.0000,2019-09-19 12:00:00,1.0000,0.0000,2019.0000
50%,7265.5000,19.5000,12.5000,4.0000,50.0000,418.0000,81.0000,14.0000,2.0000,0.0000,1.0000,5.0000,2021-04-12 12:00:00,1.0000,1.0000,2021.0000
75%,7835.2500,29.0000,16.0000,6.0000,58.0000,492.0000,85.0000,17.0000,3.0000,0.0000,2.0000,7.0000,2022-07-26 06:00:00,2.0000,1.0000,2022.0000
max,8405.0000,38.0000,37.0000,18.0000,80.0000,848.0000,93.0000,31.0000,10.0000,3.0000,9.0000,23.0000,2023-12-06 00:00:00,7.0000,6.0000,2023.0000
std,658.2514,10.9671,5.2767,2.3665,11.0123,110.9565,5.9931,4.5684,1.4466,0.3734,1.3782,3.0315,NaN,1.1503,0.9844,1.7073


In [19]:
#Passa para o campeonado de 2020 todos os jogos anteriores a Maio/2021.
filtro_pandemia = (df_limpeza["data"].dt.year == 2021) & (df_limpeza["data"] < "2021-05-01 00:00:00")

df_limpeza.loc[filtro_pandemia, "ano"] = 2020

jogos_por_ano = df_limpeza.groupby("ano").size().reset_index(name="quantidade de linhas por campeonato")
print(jogos_por_ano)

    ano  quantidade de linhas por campeonato
0  2018                                  760
1  2019                                  760
2  2020                                  760
3  2021                                  760
4  2022                                  760
5  2023                                  760


In [20]:
# Mapeia as linhas que possuem qualquer NaN
filtro = df_limpeza.isna().any(axis=1)
print(f"Linhas problemáticas identificadas: {filtro.sum()}")

# Lista completa de colunas que devem ser anuladas nessas linhas
colunas_para_preencher_nan = ['chutes', 'chutes_no_alvo', 'posse_de_bola', 'passes', 'precisao_passes', 'faltas', 'cartao_amarelo', 'cartao_vermelho', 'impedimentos', 'escanteios']

# 2. Atribui pd.NA diretamente para TODAS as colunas selecionadas APENAS nessas 48 linhas
df_limpeza.loc[filtro, colunas_para_preencher_nan] = pd.NA

print("Conversão concluída! Todos os campos dessas colunas nas 48 linhas agora são <NA>.")



Linhas problemáticas identificadas: 48
Conversão concluída! Todos os campos dessas colunas nas 48 linhas agora são <NA>.


In [21]:
df_limpeza.isna().any(axis=1).sum()

np.int64(48)

In [22]:
# Mapeia as linhas que possuem qualquer NaN
filtro = df_limpeza.isna().any(axis=1)

colunas_mediana = ['chutes', 'chutes_no_alvo', 'passes', 'precisao_passes', 'faltas', 'escanteios']

for col in colunas_mediana:
    # Registra o total de linhas com NaN e a média antes
    linhas_com_nan_antes = df_limpeza.isna().any(axis=1).sum()
    media_antes = df_limpeza[col].mean()
    
    # 2. Calcula a mediana normalmente baseada no grupo (clube e ano)
    mediana_grupo = df_limpeza.groupby(['clube', 'ano'])[col].transform('median').astype(float)
    
    # 3. Usa o .loc para garantir que o fillna atue APENAS nas 48 linhas filtradas
    df_limpeza.loc[filtro, col] = df_limpeza.loc[filtro, col].fillna(mediana_grupo)
    
    # Registra o total de linhas com NaN após o preenchimento desta coluna
    linhas_com_nan_depois = df_limpeza.isna().any(axis=1).sum()
    
    print(f"--- Coluna: {col} ---")
    print(f"Linhas com NaN antes  : {linhas_com_nan_antes}")
    print(f"Original              : {len(df_limpeza)} linhas | média {media_antes:.4f}")
    print(f"Linhas com NaN depois : {linhas_com_nan_depois}")
    print(f"Ajustado              : {len(df_limpeza)} linhas | média {df_limpeza[col].mean():.4f}\n")

print("Total final de linhas com qualquer NAN restante:", df_limpeza.isna().any(axis=1).sum())





--- Coluna: chutes ---
Linhas com NaN antes  : 48
Original              : 4560 linhas | média 13.1031
Linhas com NaN depois : 48
Ajustado              : 4560 linhas | média 13.1042

--- Coluna: chutes_no_alvo ---
Linhas com NaN antes  : 48
Original              : 4560 linhas | média 4.2819
Linhas com NaN depois : 48
Ajustado              : 4560 linhas | média 4.2838

--- Coluna: passes ---
Linhas com NaN antes  : 48
Original              : 4560 linhas | média 424.6152
Linhas com NaN depois : 48
Ajustado              : 4560 linhas | média 424.6781

--- Coluna: precisao_passes ---
Linhas com NaN antes  : 48
Original              : 4560 linhas | média 80.4337
Linhas com NaN depois : 48
Ajustado              : 4560 linhas | média 80.4493

--- Coluna: faltas ---
Linhas com NaN antes  : 48
Original              : 4560 linhas | média 14.7708
Linhas com NaN depois : 48
Ajustado              : 4560 linhas | média 14.7743

--- Coluna: escanteios ---
Linhas com NaN antes  : 48
Original           

In [23]:
#coluna para preencher 50%
df_limpeza['posse_de_bola'] = df_limpeza['posse_de_bola'].fillna(50)

In [24]:
# Exibe apenas as linhas que possuem algum valor NaN na tabela
print(df_limpeza.isna().any(axis=1).sum())
display(df_limpeza[df_limpeza.isna().any(axis=1)])


48


,partida_id,rodada,clube,chutes,chutes_no_alvo,posse_de_bola,passes,precisao_passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,data,hora,mandante,visitante,formacao_mandante,formacao_visitante,vencedor,mandante_Placar,visitante_Placar,mandante_Estado,visitante_Estado,ano
13030,6517,2,CSA,10.0,3.0,50.0,344.0,79.0,16.0,NaN,NaN,NaN,3.0,2019-05-01,16:00,CSA,Palmeiras,4-2-3-1,4-2-3-1,-,1,1,AL,SP,2019
13031,6517,2,Palmeiras,15.0,6.0,50.0,385.0,80.5,15.5,NaN,NaN,NaN,5.0,2019-05-01,16:00,CSA,Palmeiras,4-2-3-1,4-2-3-1,-,1,1,AL,SP,2019
13076,6541,4,Atletico-MG,15.0,5.0,50.0,419.0,81.0,14.0,NaN,NaN,NaN,5.0,2019-05-12,10:53,Atletico-MG,Palmeiras,4-3-3,4-2-3-1,Palmeiras,0,2,MG,SP,2019
13077,6541,4,Palmeiras,15.0,6.0,50.0,385.0,80.5,15.5,NaN,NaN,NaN,5.0,2019-05-12,10:53,Atletico-MG,Palmeiras,4-3-3,4-2-3-1,Palmeiras,0,2,MG,SP,2019
13178,6595,9,Athletico-PR,13.5,5.0,50.0,444.5,82.5,16.0,NaN,NaN,NaN,6.0,2019-06-13,14:08,Goias,Athletico-PR,4-2-3-1,4-1-2-1-2,Goias,2,1,GO,PR,2019
13179,6595,9,Goias,13.0,4.0,50.0,325.5,79.0,14.0,NaN,NaN,NaN,5.0,2019-06-13,14:08,Goias,Athletico-PR,4-2-3-1,4-1-2-1-2,Goias,2,1,GO,PR,2019
13210,6609,11,Athletico-PR,13.5,5.0,50.0,444.5,82.5,16.0,NaN,NaN,NaN,6.0,2019-07-20,16:10,CSA,Athletico-PR,4-3-1-2,4-2-3-1,Athletico-PR,0,4,AL,PR,2019
13211,6609,11,CSA,10.0,3.0,50.0,344.0,79.0,16.0,NaN,NaN,NaN,3.0,2019-07-20,16:10,CSA,Athletico-PR,4-3-1-2,4-2-3-1,Athletico-PR,0,4,AL,PR,2019
13228,6617,12,Cruzeiro,14.0,4.0,50.0,454.0,82.0,14.0,NaN,NaN,NaN,4.0,2019-07-27,13:58,Cruzeiro,Athletico-PR,4-2-3-1,4-2-3-1,Athletico-PR,0,2,MG,PR,2019
13229,6617,12,Athletico-PR,13.5,5.0,50.0,444.5,82.5,16.0,NaN,NaN,NaN,6.0,2019-07-27,13:58,Cruzeiro,Athletico-PR,4-2-3-1,4-2-3-1,Athletico-PR,0,2,MG,PR,2019


Cartao_amarelo, cartao_vermelho e impedimentos serão mantidos com NaN porque são valores baixos e qualquer variação pode impactar nas estatísticas.


In [34]:
import numpy as np

# Geração da coluna total_gols'
df_limpeza["total_gols"] = df_limpeza["mandante_Placar"]+df_limpeza["visitante_Placar"]

# Geração da coluna resultado comparando o valor das colunas vencedor e clube
df_limpeza["resultado"] = np.select([df_limpeza["vencedor"] == "-", df_limpeza["vencedor"] == df_limpeza["clube"]], ['Empate', 'Vitória'], default='Derrota')

# Geração da coluna pontos mapeando os valores de 'resultado'
df_limpeza["pontos"] = df_limpeza["resultado"].map({'Vitória': 3, 'Empate': 1, 'Derrota': 0})


Foram criadas as colunas ano, na etapa de diagnóstico, total_gols, resultado e pontos, na etapa de limpeza.

In [35]:
coluna = 'total_gols'

# 1. Calcula a média e desvio padrão do ano
media_ano = df_limpeza.groupby(['ano'])[coluna].transform('mean')
std_ano = df_limpeza.groupby(['ano'])[coluna].transform('std')

# Salva a média calculada em uma nova coluna para exibição
df_limpeza['media_gols_ano'] = media_ano

# 2. Evita divisão por zero caso o desvio seja 0
std_ano = std_ano.replace(0, 1)

# 3. Calcula o Z-Score
df_limpeza['z_score_' + coluna] = (df_limpeza[coluna] - media_ano) / std_ano

# 4. Filtra outliers extremos E garante apenas uma linha por partida (clube == mandante)
condicao_outlier = df_limpeza['z_score_' + coluna].abs() > 3.5
condicao_mandante = df_limpeza['clube'] == df_limpeza['mandante']

outliers_jogo = df_limpeza[condicao_outlier & condicao_mandante]

print(f"Partidas que foram outliers de {coluna} no ano (incluindo a média anual):")

# 5. Exibe a tabela incluindo a nova coluna formatada com 2 casas decimais
colunas_exibicao = ['ano', 'data', 'mandante', 'mandante_Placar', 'visitante', 'visitante_Placar', coluna, 'media_gols_ano', 'z_score_' + coluna]
display(outliers_jogo[colunas_exibicao].sort_values(by='data'))



Partidas que foram outliers de total_gols no ano (incluindo a média anual):


,ano,data,mandante,mandante_Placar,visitante,visitante_Placar,total_gols,media_gols_ano,z_score_total_gols
13066,2019,2019-05-05,Gremio,4,Fluminense,5,9,2.305263,4.298396
13418,2019,2019-09-26,Palmeiras,6,CSA,2,8,2.305263,3.656340
13650,2019,2019-11-13,Flamengo,4,Vasco,4,8,2.305263,3.656340
13889,2020,2020-09-02,Bahia,3,Flamengo,5,8,2.484211,3.575959
14072,2020,2020-10-14,Sport,3,Internacional,5,8,2.484211,3.575959
15480,2022,2022-06-08,Fluminense,5,Atletico-MG,3,8,2.381579,3.837035
16550,2023,2023-10-07,Goias,4,Bahia,6,10,2.489474,4.601609


ezdesvio padrão e score z

In [38]:
# 1. Agrupa por ano e por clube para calcular as estatísticas por time
estudo_violencia = df_limpeza.groupby(['ano', 'clube'])['faltas'].agg(
    total_faltas='sum',
    media_faltas_por_jogo='mean',
    desvio_padrao_faltas='std',
).reset_index()

# 2. CORREÇÃO: Calcula a média de faltas POR JOGO de todos os times do ano
estudo_violencia['media_faltas_por_jogo_ano'] = estudo_violencia.groupby('ano')['media_faltas_por_jogo'].transform('mean')

# 3. Ordena para garantir que o time com mais faltas fique no topo do seu ano
estudo_violencia = estudo_violencia.sort_values(by=['ano', 'total_faltas'], ascending=[False, False])

# 4. Isola apenas o primeiro registro (o maior) de cada ano
maiores_por_ano = estudo_violencia.groupby('ano').first().reset_index()
menores_por_ano = estudo_violencia.groupby('ano').last().reset_index()

print("📊 O time mais violento de cada ano comparado à média de faltas por jogo da temporada:")
display(maiores_por_ano)

print("📊 O time menos violento de cada ano comparado à média de faltas por jogo da temporada:")
display(menores_por_ano)


📊 O time mais violento de cada ano comparado à média de faltas por jogo da temporada:


,ano,clube,total_faltas,media_faltas_por_jogo,desvio_padrao_faltas,media_faltas_por_jogo_ano
0,2018,Bahia,675.0,17.763158,5.206203,15.553947
1,2019,Sao Paulo,639.0,16.815789,5.612993,14.703947
2,2020,Internacional,666.0,17.526316,4.397400,15.648684
3,2021,Internacional,660.0,17.368421,3.340344,14.961842
4,2022,America-MG,577.0,15.184211,4.887133,13.440789
5,2023,Goias,637.0,16.763158,3.851530,14.336842


📊 O time menos violento de cada ano comparado à média de faltas por jogo da temporada:


,ano,clube,total_faltas,media_faltas_por_jogo,desvio_padrao_faltas,media_faltas_por_jogo_ano
0,2018,Gremio,529.0,13.921053,3.641960,15.553947
1,2019,Botafogo-RJ,474.0,12.473684,4.452372,14.703947
2,2020,Gremio,511.0,13.447368,3.902164,15.648684
3,2021,Corinthians,419.0,11.026316,4.426498,14.961842
4,2022,Fluminense,431.0,11.342105,3.265514,13.440789
5,2023,Cuiaba,473.0,12.447368,3.500559,14.336842


O clube que teve mais faltas por 2 anos consecutivos foi o Internacional.
<br>
O clube com menos faltas em 2 anos não consecutivos foi o Grêmio.

In [28]:
#No final salva o arquivo tratado em .csv para poder realizar as análises sem precisar retornar às etapas anteriores
df_limpeza.to_csv('campeonato_brasileiro_tratado.csv', encoding='utf-8', index=False)

<font size='5'>4 Análise exploratória</font>
<br>
<br>Respostas às perguntas definidas no item 1, com evidência numérica
<br>Uso obrigatório de: `groupby` com agregação múltipla, ao menos um `merge` (pode ser com uma tabela auxiliar que vocês montarem) e ao menos uma `pivot_table`
<br>Ao menos 4 gráficos, cada um acompanhado de uma leitura em texto (o gráfico sozinho não conta)
<br>Uso de NumPy em pelo menos um cálculo relevante (normalização, z-score, operação matricial, simulação)

In [29]:
#1 Abrir arquivo CSV com o relatório tratatado
import pandas as pd

df = pd.read_csv('campeonato_brasileiro_tratado.csv')

df.head()

,partida_id,rodada,clube,chutes,chutes_no_alvo,posse_de_bola,passes,precisao_passes,faltas,cartao_amarelo,cartao_vermelho,impedimentos,escanteios,data,hora,mandante,visitante,formacao_mandante,formacao_visitante,vencedor,mandante_Placar,visitante_Placar,mandante_Estado,visitante_Estado,ano,total_gols,resultado,pontos,media_gols_ano,z_score_total_gols
0,6126,1,Cruzeiro,12.0,2.0,40.0,425.0,81.0,20.0,3.0,0.0,0.0,7.0,2018-04-14,16:00,Cruzeiro,Gremio,4-2-3-1,4-2-3-1,Gremio,0,1,MG,RS,2018,1,Derrota,0,2.176316,-0.792936
1,6126,1,Gremio,6.0,2.0,60.0,649.0,86.0,15.0,1.0,1.0,3.0,4.0,2018-04-14,16:00,Cruzeiro,Gremio,4-2-3-1,4-2-3-1,Gremio,0,1,MG,RS,2018,1,Vitória,3,2.176316,-0.792936
2,6127,1,Vitoria,21.0,6.0,66.0,545.0,88.0,24.0,3.0,0.0,2.0,4.0,2018-04-14,19:02,Vitoria,Flamengo,4-2-3-1,4-1-4-1,-,2,2,BA,RJ,2018,4,Empate,1,2.176316,1.229316
3,6127,1,Flamengo,10.0,5.0,34.0,286.0,72.0,10.0,2.0,1.0,2.0,5.0,2018-04-14,19:02,Vitoria,Flamengo,4-2-3-1,4-1-4-1,-,2,2,BA,RJ,2018,4,Empate,1,2.176316,1.229316
4,6128,1,Santos,22.0,3.0,54.0,532.0,90.0,7.0,0.0,0.0,1.0,4.0,2018-04-14,21:03,Santos,Ceara,4-3-3,4-1-4-1,Santos,2,0,SP,CE,2018,2,Vitória,3,2.176316,-0.118852


In [30]:
# Criar coluna de gols do clube em cada linha
def get_gols(row):
    if row['clube'] == row['mandante']:
        return row['mandante_Placar']
    else:
        return row['visitante_Placar']

df['gols'] = df.apply(get_gols, axis=1)

pd.set_option('display.max_columns', None)
df['data'] = pd.to_datetime(df['data'])
df['ano'] = df['data'].dt.year

def get_resultado(row):
    if row['vencedor'] == row['clube']:
        return 'V'
    elif row['vencedor'] == '-':
        return 'E'
    else:
        return 'D'
df['resultado'] = df.apply(get_resultado, axis=1)
df['pontos'] = df['resultado'].map({'V': 3, 'E': 1, 'D': 0})


<center><h3>Analise de Eficiência Ofensiva</h3></center>




<font size="3">
Para analisar a eficiência ofensiva foram excluídos os casos onde chutes é igual a 0, pois a eficiência ofensiva não pode ser calculada nesses casos. Além disso, foram excluídos os casos onde chutes é menor que gols, pois isso indica um erro nos dados (não é possível marcar mais gols do que chutes).
Filtragem foi feita com base nessas condições, e o shape do dataframe antes e depois da filtragem foi impresso para referência.

 </font>

In [31]:
df_eficiencia = df[['partida_id', 'data', 'ano', 'clube', 'gols', 'chutes', 'resultado']].copy()
print(f"Linhas com chute = 0: {df_eficiencia[df_eficiencia['chutes'] == 0].shape[0]}")
print(f"Shape antes da filtragem: {df_eficiencia.shape}")
df_eficiencia = df_eficiencia[df_eficiencia['chutes'] > 0]
print(f"Shape após a filtragem: {df_eficiencia.shape}")
df_eficiencia['eficiencia_partida'] = df_eficiencia['gols'] / df_eficiencia['chutes']

Linhas com chute = 0: 0
Shape antes da filtragem: (4560, 7)
Shape após a filtragem: (4560, 7)


In [32]:
eficiencia_agrupada = (
    df.groupby(["clube", "ano"])[["gols", "chutes"]]
    .sum()
    .reset_index()
)

eficiencia_agrupada['eficiencia'] = (eficiencia['gols'] / eficiencia['chutes']) * 100

eficiencia = eficiencia.sort_values(
    by=["ano", "eficiencia"], ascending=[True, False]
)

eficiencia

NameError: name 'eficiencia' is not defined

In [ ]:
# 6. Criar o ranking dentro de cada temporada
# O 'ascending=False' garante que a maior eficiência ganhe o 1º lugar.
# O 'method='min'' serve para tratar empates de forma justa.
eficiencia["ranking_ofensivo"] = (
    eficiencia.groupby("ano")["eficiencia"]
    .rank(ascending=False, method="dense") #optamos pelo uso de dense para que em caso de empate o próximo ranking seja o próximo número inteiro, sem pular números.
    .astype(int)
)

# 7. Ordenar visualmente o DataFrame por temporada e depois pelo ranking
eficiencia = eficiencia.sort_values(
    by=["ano", "ranking_ofensivo"]
)

# Visualizar o resultado na tela
print(eficiencia[["ano", "ranking_ofensivo", "clube", "eficiencia"]])


In [ ]:
pontuacao = ( df.groupby(["clube", "ano"])["pontos"].sum().reset_index().sort_values(by=["ano", "pontos"], ascending=[True, False]
    )  
)

pontuacao

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# 1. Identificar as temporadas únicas nos dados
temporadas = eficiencia["ano"].unique()
num_temporadas = len(temporadas)

# 2. Configurar a área do gráfico dinamicamente com base no número de anos
fig, axes = plt.subplots(
    nrows=num_temporadas,
    ncols=1,
    figsize=(16, 6 * num_temporadas),
    gridspec_kw={"right": 0.75},
)

# Garantir que 'axes' seja sempre uma lista, mesmo se houver apenas 1 temporada
if num_temporadas == 1:
    axes = [axes]

# 3. Criar um gráfico de barras para cada temporada
for idx, ano in enumerate(temporadas):
    # Filtrar os dados de eficiência apenas do ano atual do loop
    dados_ano = eficiencia[eficiencia["ano"] == ano].copy()  # .copy() evita avisos do pandas

    # Criar a barra horizontal usando o Seaborn
    ax = axes[idx]
    sns.barplot(
        x="eficiencia",
        y="clube",
        data=dados_ano,
        palette="viridis",
        ax=ax,
        hue="clube",
        legend=False,
    )

    ## ------------------ ALTERAÇÃO AQUI: RANKING À ESQUERDA ------------------ ##
    # Criamos uma lista de textos combinando o Ranking + Nome do Clube
    novos_rotulos = [
        f"{row['ranking_ofensivo']}º - {row['clube']}" 
        for _, row in dados_ano.iterrows()
    ]
    # Substituímos os nomes simples do eixo Y por essa nova lista formatada
    ax.set_yticklabels(novos_rotulos)
    ## ----------------------------------------------------------------------- ##

    # Adicionar APENAS o valor da eficiência ao lado de cada barra
    for container in ax.containers:
        for bar in container:
            width = bar.get_width()
            if width > 0:
                y_pos = bar.get_y() + bar.get_height() / 2

                # Ajustado para mostrar apenas a porcentagem, já que o ranking foi para a esquerda
                texto = f"{width:.2f}"

                # Coloca o texto imediatamente após o fim da barra
                ax.text(
                    width + 0.005,
                    y_pos,
                    texto,
                    va="center",
                    ha="left",
                    fontsize=9,                )

    # --- SOLUÇÃO DEFINITIVA PARA NÃO CORTAR TEXTO ---
    ax.margins(x=0.15)

    # Garante que os números da régua (ticks) fiquem visíveis neste subgráfico
    ax.tick_params(axis="x", labelbottom=True)

    # --- Buscar o Top 3 do campeonato no DF de pontuação ---
    pontos_ano = pontuacao[pontuacao["ano"] == ano].head(3)

    time_campeao = pontos_ano.iloc[0]["clube"]
    time_vice = pontos_ano.iloc[1]["clube"]
    time_terceiro = pontos_ano.iloc[2]["clube"]

    # Monta a caixinha de texto formatada
    texto_podio = (
        f" PÓDIO DO CAMPEONATO\n"
        f"--------------------------------\n"
        f"Campeão: {time_campeao}\n"
        f"Vice-campeão: {time_vice}\n"
        f"3º Lugar: {time_terceiro}"
    )

    # Adiciona o texto fora do gráfico (à direita)
    ax.text(
        1.05,
        0.5,
        texto_podio,
        transform=ax.transAxes,
        fontsize=11,
        verticalalignment="center",
        bbox=dict(
            boxstyle="round,pad=0.6",
            facecolor="#fdfdfd",
            edgecolor="#cccccc",
            alpha=1.0,
        ),
    )

    # Customizações estéticas por subgráfico
    ax.set_title(
        f"Ranking de Eficiência Ofensiva - Temporada {ano}", fontsize=14, pad=10
    )
    
    # Mudamos o texto do eixo X, já que o ranking mudou de lugar
    ax.set_xlabel("Eficiência (%)", fontsize=11, labelpad=12)
    ax.xaxis.get_label().set_visible(True)
    
    ax.set_ylabel("Ranking Eficiência - Clube", fontsize=11)
    ax.grid(axis="x", linestyle="--", alpha=0.5)

# Controla o espaçamento manualmente para que o nome do eixo X nunca seja cortado no fundo
plt.subplots_adjust(bottom=0.1, hspace=0.4)
plt.show()


In [ ]:
df_eficiencia



<span style="color: red; font-size: 18px"> OBS.: Aguardar limpeza dos dados para fazer a relação entre eficiencia e melhor colocação, se houver</span>
<br>
<br>
<br>
<br>

<font siza='5'>5. Conclusões</font>
<br>
<br>3 a 5 achados defensáveis, escritos para uma **área de negócio**, não para outro programador
<br>Limitações da análise: o que os dados **não** permitem afirmar
<br>Próximos passos: o que vocês investigariam com mais tempo ou mais dados

(Espaço para responder as conclusões)

--------------------------------------------------------------------------------------------------------
<font size='5'>Observações:</font>

Requisitos Git
- [ ] Repositório público, com README preenchido
- [ ] **Todos** os integrantes com commits (o histórico mostra quem fez o quê)
- [ ] No mínimo 15 commits, com mensagens descritivas (`feat: adiciona tratamento de outliers`, não `update`)
- [ ] Uso de pelo menos uma branch além da `main`
- [ ] `.gitignore` adequado (nada de `.ipynb_checkpoints`, `__pycache__` ou credenciais)
- [ ] Notebook commitado **com as saídas executadas**


Para fazer depois que o trabalho estiver pronto:
| 1 min | A base e por que ela foi escolhida |
| 2 min | Estado inicial: o que estava errado nos dados |
| 2 min | Principais decisões de tratamento e suas justificativas |
| 4 min | Os achados, com os gráficos |
| 1 min | Limitações e próximos passos |

Critérios de avaliação por rubrica

| Eixo | Em desenvolvimento | Desenvolvido | Referência |
|---|---|---|---|
| **Git** | Commits concentrados em uma pessoa ou mensagens genéricas | Histórico distribuído, mensagens claras, `.gitignore` correto | Uso fluente de branches; README que permite outra pessoa reproduzir a análise |
| **NumPy** | Uso apenas incidental | Aplica vetorização em pelo menos um cálculo relevante | Escolhe NumPy conscientemente por eficiência e explica o porquê |
| **Pandas** | Operações básicas com erros de alinhamento ou perda silenciosa de linhas | Domina `groupby`, `merge` e `pivot` e confere o resultado | Encadeia operações de forma legível e verifica integridade a cada etapa |
| **Leitura e gravação** | Só lê csv simples | Lida com configurações não triviais e grava em formato adequado | Justifica a escolha do formato de saída (ex.: parquet por tipos e tamanho) |
| **Limpeza e transformação** | Apaga o que incomoda, sem justificar | Trata faltantes, duplicados e outliers justificando cada decisão | Compara estratégias alternativas e mede o impacto de cada uma antes de decidir |
| **Comunicação** | Apresenta código, não conclusões | Achados claros, apoiados em números | Narrativa que uma área de negócio entenderia e usaria para decidir |

## Erros que mais custam nota

1. Concluir coisas que os dados não sustentam (correlação apresentada como causa).
2. Apagar linhas com faltantes sem verificar quantas e quais são — e sem dizer nada sobre isso.
3. Tratar como outlier tudo que é extremo na coluna inteira, sem olhar o grupo de comparação.
4. Fazer `merge` e não conferir se o número de linhas mudou.
5. Gráfico sem título, sem eixo nomeado e sem uma frase dizendo o que ele mostra.
6. Repositório com um único commit no dia da entrega.
